### PR TITLE
mtmd : add vision support for llama 4

### PR DESCRIPTION
**Pull request by ngxson**
> [!IMPORTANT]
>
> The implementation in this PR has been confirmed to match the logits produced by `transformers` library
>
> If there are problems regarding quality, it is likely because of the model itself, not because of llama.cpp
>
> My test code: https://gist.github.com/ngxson/ba732a8674111a558795a35afda98bdd

Pre-quantized model: https://huggingface.co/ggml-org/Llama-4-Scout-17B-16E-Instruct-GGUF

---

Perceived image is sometimes incorrect (not due to my implementation, see below)

For text model, I'm testing with `Llama-4-Scout-17B-16E-Instruct-UD-IQ2_M` quant from @UnslothAI

`llama-mtmd-cli -m ../models/Llama-4-Scout-17B-16E-Instruct-UD-IQ2_M.gguf --mmproj ../models/Llama-4-Scout-17B-16E-Instruct/mmproj-model.gguf --image ../models/lenna.png -p "How many people do you see? What is the composition of this image? If it is repeated, which direction and how many repetitions?"`

`There is one person in this image. The composition of this image is a woman wearing a hat. If this image is repeated, it is repeated horizontally and there are 2 repetitions.`

---

Ok, it sounds like a joke but I spent another day debugging this, just to discover that the model response this way on every single platform, not just llama.cpp

image

---

My thought on this (NOTE: this is purely subjective - this PR is my personal work that I do in my free time):

The [official implementation](https://github.com/meta-llama/llama-models/blob/main/models/llama4/vision/encoder.py) make me feel like this whole vision support is a last-minute addition. It's overly complicated, here is why:
1. They could use either learned embedding **or** 2D-RoPE, but they decided to use **both**, which makes the learned embedding tensor pretty much redundant. Indeed, when I remove it from crgaph, nothing changes
2. I suppose they use 2D-RoPE to support dynamic resolution like Pixtral. But no, they fix the image size to 336x336 which make 2D-RoPE useless. Indeed, the way they support dynamic resolution is to rely on preprocessor, which is something llava-uhd and minicpm-v already done (and they did a better job tbh)
3. The projector is pretty hacky, they use **both** MLP and single-matrix projector (plus not one, but **two** GELU activation in-between). Other models only use either MLP or single-matrix, not both. Otherwise, they could simply add a up/gate/down FFN and it should does the job just fine.

### CODE REVIEW

**Comment by ngxson** [comment]
Test result:

```
OK:   llama-mtmd-cli ggml-org/SmolVLM-500M-Instruct-GGUF:Q8_0
OK:   llama-mtmd-cli ggml-org/SmolVLM2-2.2B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/SmolVLM2-500M-Video-Instruct-GGUF:Q8_0
OK:   llama-mtmd-cli ggml-org/gemma-3-4b-it-GGUF:Q4_K_M
OK:   llama-mtmd-cli THUDM/glm-edge-v-5b-gguf:Q4_K_M
OK:   llama-mtmd-cli second-state/Llava-v1.5-7B-GGUF:Q2_K
OK:   llama-mtmd-cli cjpais/llava-1.6-mistral-7b-gguf:Q3_K_M
OK:   llama-mtmd-cli ibm-research/granite-vision-3.2-2b-GGUF:Q4_K_M
OK:   llama-mtmd-cli second-state/MiniCPM-Llama3-V-2_5-GGUF:Q2_K
OK:   llama-mtmd-cli openbmb/MiniCPM-V-2_6-gguf:Q2_K
OK:   llama-mtmd-cli openbmb/MiniCPM-o-2_6-gguf:Q4_0
OK:   llama-mtmd-cli bartowski/Qwen2-VL-2B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/Qwen2.5-VL-3B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/InternVL2_5-1B-GGUF:Q8_0
OK:   llama-mtmd-cli ggml-org/InternVL3-1B-Instruct-GGUF:Q8_0
OK:   llama-mtmd-cli ggml-org/pixtral-12b-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/Mistral-Small-3.1-24B-Instruct-2503-GGUF
OK:   llama-mtmd-cli ggml-org/Qwen2-VL-2B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/Qwen2-VL-7B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/Qwen2.5-VL-3B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/Qwen2.5-VL-7B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/InternVL3-8B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/InternVL3-14B-Instruct-GGUF:Q4_K_M
OK:   llama-mtmd-cli ggml-org/Llama-4-Scout-17B-16E-Instruct-GGUF:IQ1_S
```

**Comment by ggerganov** [comment]
> Ok, it sounds like a joke but I spent another day debugging this, just to discover that the model response this way on every single platform, not just llama.cpp

Huh interesting. It could also be a problem in the `transformers` implementation?

File: tools/mtmd/clip.cpp [hunk 1]
```diff
@@ -936,6 +939,98 @@ struct clip_graph {
...
+            cb(inp, "patch_conv", -1);
+        }
+
+        // add CLS token
+        inp = ggml_concat(ctx0, inp, model.class_embedding, 1);
+
+        // build ViT with 2D position embeddings
+        auto add_pos = [&](ggml_tensor * cur, const clip_layer &) {
+            return build_rope_2d(ctx0, cur, pos_w, pos_h, hparams.rope_theta, false);
+        };
```
Comment (ggerganov) [inline tools/mtmd/clip.cpp:975]: Here

**Comment by ngxson** [comment]
> Huh interesting. It could also be a problem in the `transformers` implementation?

I was confirmed from an internal source that some providers on inference provider (that I tested above) use their own implementation instead of `transformers`, so at this point it is very likely because of the model itself.

File: tools/mtmd/clip.cpp [hunk 2]
```diff
@@ -522,7 +525,7 @@ struct clip_graph {
         ggml_set_input(pos_w);
 
         auto add_pos = [&](ggml_tensor * cur, const clip_layer &) {
-            return build_rope_2d(ctx0, cur, pos_h, pos_w, hparams.rope_theta);
+            return build_rope_2d(ctx0, cur, pos_h, pos_w, hparams.rope_theta, true);
         };
```
[the remark this answers is not in the document]
Comment (ngxson) [reply tools/mtmd/clip.cpp:529]: For pixtral, the RoPE order is [H then W](https://github.com/Blaizzy/mlx-vlm/blob/a57156aa87b33cca6e5ee6cfc14dd4ef8f611be6/mlx_vlm/models/pixtral/vision.py#L219-L226), and for llama 4 it's reversed, as seen in [transformers](https://github.com/huggingface/transformers/blob/40a493c7ed4f19f08eadb0639cf26d49bfa5e180/src/transformers/models/llama4/modeling_llama4.py#L1312) and [mlx-vlm](https://github.com/Blaizzy/mlx-vlm/blob/a57156aa87b33cca6e5ee6cfc14dd4ef8f611be6/mlx_vlm/models/llama4/vision.py#L441)

For llama 4, if we reverse the order to [H, W], it will output completely incoherent result. But yeah I think it's better to add a comment here.

**Review by ggerganov — APPROVED** [review]
If the model is so broken, maybe print a warning in the logs and reference the findings in this PR.

**Comment by ngxson** [comment]
Yes that's a good idea, I added the message in https://github.com/ggml-org/llama.cpp/pull/13282/commits/d5e50aa09b1eeb0525ff1dffe31523aefa2c29de

**Comment by strawberrymelonpanda** [comment]
> Two of my Llama 4 hybrid layer quants handle this image with no issue. Strangely the largest one gets uncertain about repetition but the smaller ones are accurate.

But the smallest one says _"reflected in a mirror. The image is not repeated; it appears to be a single reflection."_, meaning it thinks there is infact repetition.  In other words, two of the three failed, rather than succeeded.

**Comment by nbergantino** [comment]
why on the git version tag b6278 (git checkout b6278) does LLama4 Scout model 4bit quantized work fine? Starting from the next tag it works like described in this conversation.

In the following the version where Llama 4 Scout works fine:
commit 34bdbbd7c2b70b848718e95bc48010f6aecd2816 (HEAD, tag: b6278)
Author: Jeff Bolz <jbolz@nvidia.com>
Date:   Mon Aug 25 23:42:44 2025 -0500

    vulkan: Remove splitting for mul_mat_id (#15568)
    
    row_ids only needs to hold the BN rows for the current tile.

In the following the last used version where Llama 4 doesn't work fine:
commit 3e18dba9fd599ace0c3f2bc94bd74e5398d2efb6 (HEAD -> master, origin/master, origin/HEAD)
Author: Jiacheng (Jason) Chen <76919340+jiachengjason@users.noreply.github.com>
Date:   Wed Nov 26 05:18:48 2025 -0500

    HIP: Patch failed testcase in WMMA-MMQ kernels for RDNA 4  (#17502)
    
    * patch failed test case MUL_MAT(type_a=q4_0,type_b=f32,m=576,n=512,k=576,bs=[1,1],nr=[1,1],per=[0,1,2,3],k_v=0,o=1) for enabling WMMA on RDNA4
    
    * Quick clean up on mma.cuh to add ggml_cuda_memcpy_1 back in for half2 and bfloat162
